### PR TITLE
OPD Profile overview & visits tab

### PR DESCRIPTION
**Pull request by ekigamba**

### CODE REVIEW

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/application/GizMalawiApplication.java [hunk 1]
```diff
@@ -142,8 +143,8 @@ public static CommonFtsObject createCommonFtsObject() {
             return names.toArray(new String[names.size()]);
         } else if (tableName.equals(DBConstantsUtils.WOMAN_TABLE_NAME)) {
             return new String[]{DBConstantsUtils.KeyUtils.BASE_ENTITY_ID, DBConstantsUtils.KeyUtils.FIRST_NAME, DBConstantsUtils.KeyUtils.LAST_NAME,
-                    DBConstantsUtils.KeyUtils.LAST_INTERACTED_WITH, DBConstantsUtils.KeyUtils.DATE_REMOVED};
-        } else if (tableName.equals(OpdDbConstants.KEY.TABLE)){
+                    DBConstantsUtils.KeyUtils.LAST_INTERACTED_WITH, OpdDbConstants.KEY.REGISTER_ID, DBConstantsUtils.KeyUtils.DATE_REMOVED};
+        } else if (tableName.equals(OpdDbConstants.KEY.TABLE)) {
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/application/GizMalawiApplication.java]: Naming here

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/configuration/GizOpdRegisterRowOptions.java [hunk 2]
```diff
@@ -33,22 +45,35 @@ public void populateClientRow(@NonNull Cursor cursor, @NonNull CommonPersonObjec
...
-            opdRegisterViewHolder.dueButton.setTextColor(Color.parseColor("#219e05"));
-            opdRegisterViewHolder.dueButton.setAllCaps(false);
-            opdRegisterViewHolder.dueButton.setBackgroundResource(R.color.transparent);
-            return;
+        Button dueButton = opdRegisterViewHolder.dueButton;
+        if (strVisitEndDate != null) {
+            try {
+                Date visitEndDate = new SimpleDateFormat(OpdConstants.DateFormat.YYYY_MM_DD_HH_MM_SS, Locale.ENGLISH).parse(strVisitEndDate);
+                if (OpdLibrary.getInstance().isPatientInTreatedState(visitEndDate)) {
+                    String treatedTime = (new SimpleDateFormat(GizConstants.DateFormat.HH_MM_AMPM, Locale.ENGLISH)).format(visitEndDate);
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/configuration/GizOpdRegisterRowOptions.java]: use date/string conversion functions in `OpdUtils`

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/configuration/GizOpdRegisterSwitcher.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import org.smartregister.opd.configuration.OpdRegisterSwitcher;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Created by Ephraim Kigamba - ekigamba@ona.io on 2019-11-20
+ */
+
+public class GizOpdRegisterSwitcher implements OpdRegisterSwitcher {
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/configuration/GizOpdRegisterSwitcher.java]: Please put the hard-coded strings in constants file for whole class

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/configuration/OpdRegisterQueryProvider.java [hunk 4]
```diff
@@ -20,52 +21,91 @@
 
     @NonNull
     @Override
-    public String getObjectIdsQuery(@Nullable String filters) {
-        if (TextUtils.isEmpty(filters)) {
+    public String getObjectIdsQuery(@Nullable String filters, @Nullable String mainCondition) {
+        Date latestValidCheckInDate = OpdLibrary.getInstance().getLatestValidCheckInDate();
+        String oneDayAgo = new SimpleDateFormat(OpdDbConstants.DATE_FORMAT, Locale.US).format(latestValidCheckInDate);
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/configuration/OpdRegisterQueryProvider.java]: Use date/string conversion functions in `OpdUtils`

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/configuration/OpdRegisterQueryProvider.java [hunk 5]
```diff
@@ -20,52 +21,91 @@
...
 
     @NonNull
     @Override
     public String mainSelectWhereIDsIn() {
-        Calendar calendar = Calendar.getInstance();
-        calendar.add(Calendar.DAY_OF_MONTH, -1);
-        String oneDayAgo = new SimpleDateFormat(OpdDbConstants.DATE_FORMAT, Locale.US).format(calendar.getTime());
-        String sqlQuery = "Select ec_child.id as _id, ec_child.first_name, ec_child.last_name, ec_child.middle_name, ec_child.gender, ec_child.dob, ec_child.home_address, 'Child' AS register_type, ec_child.relational_id AS relationalid, ec_child.zeir_id AS register_id, ec_child.last_interacted_with, ec_mother.first_name AS mother_first_name, ec_mother.last_name AS mother_last_name, ec_mother.middle_name AS mother_middle_name, opd_details.current_visit_start_date >= '$latest_start_visit_date' AS pending_diagnose_and_treat,  'ec_child' as entity_table, opd_details.current_visit_end_date FROM ec_child \n" +
+        Date latestValidCheckInDate = OpdLibrary.getInstance().getLatestValidCheckInDate();
+        String oneDayAgo = new SimpleDateFormat(OpdDbConstants.DATE_FORMAT, Locale.US).format(latestValidCheckInDate);
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/configuration/OpdRegisterQueryProvider.java]: Use date/string conversion functions in `OpdUtils`

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/processor/GizMalawiProcessorForJava.java [hunk 6]
```diff
@@ -141,8 +141,13 @@ public synchronized void processClient(List<EventClient> eventClients) throws Ex
...
                         .equals(Constants.EventType.UPDATE_BITRH_REGISTRATION) || eventType
                         .equals(Constants.EventType.NEW_WOMAN_REGISTRATION) || eventType.equals(OpdConstants.EventType.OPD_REGISTRATION)) {
+
+                    if (eventType.equals(OpdConstants.EventType.OPD_REGISTRATION)) {
+                        Timber.i("This is an OPD Registration");
+                    }
+
                     if (eventType.equals(OpdConstants.EventType.OPD_REGISTRATION) && eventClient.getClient() == null) {
-                        Timber.e(new Exception(), "Cannot find client corresponding to %s with base-entity-id %d", OpdConstants.EventType.OPD_REGISTRATION, event.getBaseEntityId());
+                        Timber.e(new Exception(), "Cannot find client corresponding to %s with base-entity-id %s", OpdConstants.EventType.OPD_REGISTRATION, event.getBaseEntityId());
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/processor/GizMalawiProcessorForJava.java]: No need for `new Exception()` here

File: README.md [hunk 7]
```diff
@@ -1,3 +1,6 @@
 # opensrp-client-giz-malawi
+[![Build Status](https://travis-ci.org/OpenSRP/opensrp-client-giz-malawi.svg?branch=master)](https://travis-ci.org/OpenSRP/opensrp-client-giz-malawi)
```
[the remark this answers is not in the document]
Comment (ekigamba) [reply README.md]: I did not get that comment? Why is this not needed?

[the remark this answers is not in the document]
Comment (bennsimon) [reply README.md]: its already there, pull master

[the remark this answers is not in the document]
Comment (ekigamba) [reply README.md]: When I check the code on that file, I only see one badge for the build. Search also gives the same result
Screen Shot 2019-12-03 at 12 08 54 PM

**Pull request merged**